In [14]:
import numpy as np
from sklearn.datasets import load_iris

# Grid Search

The majority of machine learning models contain parameters that can be adjusted to vary how the model learns. For example, the logistic regression model, from **sklearn**, has a parameter `C` that controls regularization, which affects the complexity of the model.

How do we pick the best value for `C`? The best value is dependent on the data used to train the model.

## How does it work?

One method is to try out different values and then pick the value that gives the best score. This technique is known as a grid search. If we had to select the values for two or more parameters, we would evaluate all combinations of the sets of values thus forming a grid of values.

Before we get into the example it is good to know what the parameter we are changing does. Higher values of C tell the model, the training data resembles real world information, place a greater weight on the training data. While lower values of C do the opposite.

## Using Default Parameters

First let's see what kind of results we can generate without a grid search using only the base parameters.

To get started we must first load in the dataset we will be working with.

In [9]:
iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [10]:
X = iris['data']
y = iris['target']

In [11]:
from sklearn.linear_model import LogisticRegression

Creating the model, setting `max_iter` to a higher value to ensure that the model finds a result.

Keep in mind the default value for `C` in a `LogisticRegression` model is `1`.

In the example below, look at the iris data set and try to train a model with varying values for `C` in logistic regression.

In [12]:
logit = LogisticRegression(max_iter=10000)
print(logit.fit(X, y))

LogisticRegression(max_iter=10000)


#### Evaluate the model

In [13]:
logit.score(X, y)

0.9733333333333334

With the default setting of `C = 1`, we achieved a score of `0.973`.

Let's see if can do any better by implementing a grid search with difference values of `0.973`.

## Implementing Grid Search

Test with range of `C` regularization values.

In [17]:
C = np.linspace(0.25, 2, 8)
C

array([0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  ])

In [24]:
# Try out different values of C.T
score = {}
logit = LogisticRegression(max_iter=10000)
for c in C:
    logit.set_params(C=c)
    logit.fit(X, y)
    score[c] = logit.score(X, y)

In [25]:
score

{0.25: 0.9666666666666667,
 0.5: 0.9666666666666667,
 0.75: 0.9733333333333334,
 1.0: 0.9733333333333334,
 1.25: 0.98,
 1.5: 0.98,
 1.75: 0.9866666666666667,
 2.0: 0.9866666666666667}

## Results Explained

We can see that the lower values of `C` performed worse than the base parameter of `1`. However, as we increased the value of `C` to `1.75` the model experienced increased accuracy.

It seems that increasing `C` beyond this amount does not help increase model accuracy.



## Note on Best Practices

We scored our logistic regression model by using the same data that was used to train it. If the model corresponds too closely to that data, it may not be great at predicting unseen data. This statistical error is known as over fitting.

To avoid being misled by the scores on the training data, we can put aside a portion of our data and use it specifically for the purpose of testing the model. Refer to the lecture on train/test splitting to avoid being misled and overfitting.